### DISCUSSION TITLE
NameError: name '__file__' is not defined when rendering .qmd document

### DISCUSSION THREAD

**Discussion by laguill** [quarto-dev/quarto-cli#11606] [Q&A]
### Issue description
I can successfully run the code using the Jupyter extension in VS Code. However, when I attempt to render the document with Quarto, I encounter the following error:

```bash
NameError: name '__file__' is not defined
```
I have the following .qmd document:
````qmd
---
title: "Data science "
subtitle: "Pandas"
format: 
  html:
    toc: true
    toc-depth: 4
    toc-expand: true
engine: jupyter
jupyter: python3
---
#### Load dataset

```{python}
import pandas as pd
from pathlib import Path

cwd = Path(__file__)
print(cwd)
data_path = cwd.parent / "data" / "Dataset.csv"

edStatsData = pd.read_csv(data_path)
```

````
### Environment Details
- `quarto version` : 1.6.39
- `OS`: ubuntu

### Additional Context
- `__file__` works in regular Python scripts, but I suspect its behavior might differ in a Jupyter or Quarto context.
- Is there a workaround for this, such as an alternative to `__file__` in Quarto?

**Comment by mcanouil** [comment]
Could you properly format your post using code blocks for code and terminal outputs? Thanks.
If your code contains code blocks, you need to enclose it using more backticks, *i.e.*, usually four ` ```` `.
See <https://quarto.org/bug-reports.html#formatting-make-githubs-markdown-work-for-us>.

**Reply by mcanouil** [reply]
Why not use the fact that `/` means root of the project (meaning there is `_quarto.yml`) or location of the input file?

**Comment by laguill** [comment]
Thank you for your answer!

I think my code should be now better formatted

I used this method because it’s what I typically use when working with standalone Python scripts outside a structured project. However, you’re right that relying on the project root or input file location might be better in Quarto.

What I don’t understand is why this works in Jupyter but not when rendering with Quarto. Does Quarto’s rendering engine modify the environment so that `__file__` isn’t defined?

Do you have any resources or insights on this difference?

**Reply by mcanouil** [reply]
> Do you have any resources or insights on this difference?

I don't. I don't have deep technical knowledge of how Jupyter / nbconvert works.

**Reply by cscheid** [reply]
Just as a statement of facts:

- we don't modify the environment
- `__file__` is not a requirement: <https://docs.python.org/3/reference/datamodel.html#module.__file__>

This can't be a Jupyter thing, because Jupyter doesn't know anything about the internal behavior of kernels. I don't know how IPython handles it, but clearly something different is happening from the way we're calling `nbconvert`

**Reply by laguill** [reply]
Should I open an issue for that ?

**Comment by s2t2** [comment]
Running into this issue as well.

Desired code, using dynamic and robust filepath compilation that will work no matter which directory the file is run from, no matter which operating system is being used:

```python
import os

DATA_DIR = os.path.join(os.path.dirname(__file__), "..", "..", "data") 
#> "`NameError: name '__file__' is not defined`"

csv_filepath = os.path.join(DATA_DIR, "my_data.csv")
df.to_csv(csv_filepath, index=False)
```

Runs into error.

Alternative attempt, based on earlier comment in this thread proposing a reference to the root directory of the repo:

```python
DATA_DIR = "/docs/data"
#> "`OSError: Cannot save file into a non-existent directory: '/docs/data'`" 
# even though that exists in the repo (its probably referencing the absolute system root filepath instead of the repo root directory).

csv_filepath = os.path.join(DATA_DIR, "my_data.csv")
df.to_csv(csv_filepath, index=False)
```

Runs into error.

Hard-coding the relative reference does work, but isn't an os agnostic solution, as windows slashes go the other way:

```python
DATA_DIR = "../../data"

csv_filepath = os.path.join(DATA_DIR, "my_data.csv")
df.to_csv(csv_filepath, index=False)
```

**Reply by gordonwoodhull** [reply]
@s2t2, surely 

```python
DATA_DIR = os.path.join("..", "..", "data")
```
 should work?

**Reply by s2t2** [reply]
Yes, that's also an option. Thank you.

Without the `__file__` anchor, I believe it may still behave differently depending on which directory you run the file from. However as long as you run the file from the same directory each time, it should be fine.

**Comment by vccortez** [comment]
I just ran into this issue.  It would be really helpful if Quarto could provide the path of the current QMD file being rendered in a way that Python or any other language could access.

Edit:
If, however, you're only interested in the folder of the current file being rendered, you can get that from the `QUARTO_DOCUMENT_PATH` environment variable.

**Comment by mcanouil** [comment]
FYI, `QUARTO_DOCUMENT_PATH` is the path to the parent directory of the document being rendered.
You also have `QUARTO_PROJECT_ROOT`.

These are documented in [Environment Variables](https://quarto.org/docs/advanced/environment-vars.html#variables-quarto-sets)

**Comment by cderv** [comment]
FYI related PR that could help 
- https://github.com/quarto-dev/quarto-cli/pull/12271

**Reply by cderv** [reply]
We've added `QUARTO_DOCUMENT_FILE` environment variable that could be use in R, Python or Julia to access the filename
